File: leetcode-algorithms/429. N-ary Tree Level Order Traversal/Solution.cpp
// Given an n-ary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).
//
// For example, given a 3-ary tree:
// We should return its level order traversal:
//
// [
//      [1],
//      [3,2,4],
//      [5,6]
// ]
//
//
// Note:
//
// The depth of the tree is at most 1000.
// The total number of nodes is at most 5000.
/*
// Definition for a Node.
class Node {
public:
    int val = NULL;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
public:
    vector<vector<int>> levelOrder(Node* root) {
        vector<vector<int>> rslt;
        if (root == nullptr) return rslt;

        queue<Node*> que;
        que.push(root);

        while (!que.empty()) {
            int size = que.size();
            vector<int> level;
            for (int i = 0; i < size; i++) {
                auto top = que.front();
                que.pop();
                level.push_back(top->val);

                for (auto c : top->children) {
                    if (c != nullptr) que.push(c);
                }
            }
            rslt.push_back(level);
        }

        return rslt;
    }
};

File: leetcode-algorithms/429. N-ary Tree Level Order Traversal/Solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public List<List<Integer>> levelOrder(Node root) {
    	List<List<Integer>> rslt = new ArrayList<>();
    	if (root == null) return rslt;

        Queue<Node> que = new LinkedList<>();
        que.add(root);

        while (!que.isEmpty()) {
        	int size = que.size();
        	List<Integer> level = new ArrayList<>();
        	for (int i = 0; i < size; i++) {
        		Node top = que.poll();
        		level.add(top.val);

        		for (Node child : top.children) {
        			if (child != null) que.add(child);
        		}
        	}
        	rslt.add(level);
        }

        return rslt;
    }
}
